### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by Sidduppal**
Merge the `documentation` with the `dev` branch. 
Includes:
- Description of each script with its functions (automatic)
- How to install and run autometa
- Automatic copyright update and last updated details
- Readme and template file for anyone willing to contribute

### CODE REVIEW

File: autometa.py [hunk 1]
```diff
@@ -137,7 +141,6 @@ def main(args):
         main(args)
     except KeyboardInterrupt:
         logger.info('User cancelled run. Exiting...')
-        sys.exit(1)
```
Comment (jason-c-kwan) [inline autometa.py:140]: Why is this line deleted?

File: autometa/common/external/bowtie.py [hunk 2]
```diff
@@ -147,6 +148,10 @@ def main(args):
         nproc=args.nproc,
         kwargs=args.kwargs)
 
+
+
+
+
```
Comment (jason-c-kwan) [inline autometa/common/external/bowtie.py:154]: Is there a reason to insert these blank lines?

File: autometa/validation/cluster_process.py [hunk 3]
```diff
@@ -1,221 +1,268 @@
...
+        taxonomy_output_path = os.path.join(args.outdir,'cluster_taxonomy.tab')
+        cmd = f'cluster_taxonomy.py -t {args.bin_table} -c {args.column} -x {args.db_dir} -o {taxonomy_output_path}'
+
+        run_command(cmd)
+
+
+def main(args):
+    logger.info(args.hello_world)
+    # operations on args.positional
+    # operations on args.optional
```
Comment (jason-c-kwan) [inline autometa/validation/cluster_process.py:240]: There are two definitions of function `main`. This second one should be taken out.

File: autometa.py (hunk 1 above)
Comment (WiscEvan) [reply autometa.py:140]: In short, It is unnecessary. At this point in the code, the error is handled and the next action will be to exit.

File: docs/source/index.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+
+Indices and tables
+------------------
+
+* :ref:`genindex`
+* :ref:`modindex`
+* :ref:`search`
+
+.. _Autometa: https://academic.oup.com/nar/article/47/10/e57/5369936
```
Comment (jason-c-kwan) [inline docs/source/index.rst:49]: This should probably be more prominent, under a section that says something like "Citing Autometa". In the old documentation, I put something like "If you find Autometa useful in your work, please cite our NAR paper [link]". Also, it might be good to have this near the top.

File: autometa/common/external/bowtie.py (hunk 2 above)
Comment (Sidduppal) [reply autometa/common/external/bowtie.py:154]: No, fixed now

File: autometa/validation/cluster_process.py (hunk 3 above)
Comment (Sidduppal) [reply autometa/validation/cluster_process.py:240]: Done!

File: docs/source/index.rst (hunk 4 above)
Comment (Sidduppal) [reply docs/source/index.rst:49]: Screenshot from 2020-04-02 14-15-32

Right now it appears just above **Quickstart** do you want a separate section for this?
`.. _Autometa: https://academic.oup.com/nar/article/47/10/e57/5369936` just link the Autometa keyword to the url of the paper.

Comment (jason-c-kwan) [reply docs/source/index.rst:49]: The placement is good, but can you please include the full citation for the paper?

File: docs/source/Scripts/Config/evnviron.rst [hunk 5]
```diff
@@ -0,0 +1,11 @@
+==========
```
[the remark this answers is not in the document]
Comment (Sidduppal) [reply docs/source/Scripts/Config/evnviron.rst:1]: What happens if I manually rename it? I guess it would become `untracked` right?

File: docs/source/index.rst [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+host-associated and highly complex ones. Autometa is copyright 2020
+Ian J. Miller, Evan R. Rees, Izaak Miller and Jason C. Kwan, and is
+released under the GNU Affero General Public License v3 (see :ref:`License`).
+
+If you find Autometa useful to your work, please cite our Autometa_ paper:
+
+Miller, I. J.; Rees, E. R.; Ross, J.; Miller, I.; Baxa, J.; Lopera, J.; 
+Kerby, R. L.; Rey, F. E.; Kwan, J. C. Autometa: Automated extraction of microbial 
+genomes from individual shotgun metagenomes. *Nucleic Acids Research*, **2019**. 
+DOI: https://doi.org/10.1093/nar/gkz148
```
[the remark this answers is not in the document]
Comment (Sidduppal) [reply docs/source/index.rst:17]: I tried it but it doesn't make a box around the text. `topic` is a body element to include some side text. The above link is the only place I could find about it.
https://docutils.sourceforge.io/docs/ref/rst/directives.html#topic

File: docs/source/contribution-guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,7 @@
+=======================
+Contribution Guidelines
+=======================
+
+Guidelines about how to contribute
+
```
[the remark this answers is not in the document]
Comment (Sidduppal) [reply docs/source/contribution-guidelines.rst:6]: I have added a fancy `todo` check it out. Also if you like we can make a separate page for To do's it can list all that stuff that we plan to do. All of it will be automatic 
https://www.sphinx-doc.org/en/master/development/tutorials/todo.html 
https://www.sphinx-doc.org/en/master/usage/extensions/todo.html

File: docs/source/conf.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+
+sys.path.insert(0, os.path.abspath('../..'))
+sys.path.insert(0, os.path.abspath('../../autometa/common'))
+sys.path.insert(0, os.path.abspath('../../autometa/config'))
+sys.path.insert(0, os.path.abspath('../../autometa/binning'))
+sys.path.insert(0, os.path.abspath('../../autometa/datasets'))
+sys.path.insert(0, os.path.abspath('../../autometa/validation'))
+sys.path.insert(0, os.path.abspath('../../autometa/taxonomy'))
+sys.path.insert(0, os.path.abspath('../../autometa/common/external'))
```
[the remark this answers is not in the document]
Comment (Sidduppal) [reply docs/source/conf.py:21]: Rejoice captain !! A script was written last night to serve this purpose. 

A side note : You make me look for stuff that I don't even think can exist, for that I'm truly grateful and happy :) 

Now rejoice !

**Comment by Sidduppal** [comment]
Some tiny details that can be missed while examining the changes:
- added sidebar in `common.rst`
- added todo flag in `contribution guidelines` and `examining results`. A note whether to make a separate page for todo is in one of the comments
- `.. topic ::` does not work

This is better and better everyday :)

File: docs/source/Scripts/Common/common.rst [hunk 9]
```diff
@@ -2,10 +2,16 @@
 Common
 ======
 
+.. sidebar:: Script under external
+
+    The external directory is a subdirectory of common.
+    All scripts in there come under common.
+
```
Comment (WiscEvan) [inline docs/source/Scripts/Common/common.rst:9]: `.. sidebar:: External Contents`

"The External directory contains wrappers respective to each utilized external program"

File: docs/source/_ext/todo.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    app.add_directive('todo', TodoDirective)
+    app.add_directive('todolist', TodolistDirective)
+    app.connect('doctree-resolved', process_todo_nodes)
+    app.connect('env-purge-doc', purge_todos)
+
+    return {
+        'version': '0.1',
+        'parallel_read_safe': True,
+        'parallel_write_safe': True,
+    }
```
Comment (WiscEvan) [inline docs/source/_ext/todo.py:127]: Another GH annotation of no newline at the end of this file.

Also, a note about the script as a whole:

Is this pulled from somewhere or did you write this? 

- If you retrieved this from somewhere you need to reference/attribute this _at least_ within this script. 
- If you did write this, some comments would be helpful to understand what is going on.

File: docs/source/conf.py [hunk 11]
```diff
@@ -11,14 +11,10 @@
...
-sys.path.insert(0, os.path.abspath('../../autometa/config'))
-sys.path.insert(0, os.path.abspath('../../autometa/binning'))
-sys.path.insert(0, os.path.abspath('../../autometa/datasets'))
-sys.path.insert(0, os.path.abspath('../../autometa/validation'))
-sys.path.insert(0, os.path.abspath('../../autometa/taxonomy'))
-sys.path.insert(0, os.path.abspath('../../autometa/common/external'))
+sys.path.append(os.path.abspath("./_ext"))
+
+for (dir_path,dir_names,file_name) in os.walk('../../', topdown=True): 
+    sys.path.insert(0, os.path.abspath(dir_path))
```
Comment (WiscEvan) [inline docs/source/conf.py:17]: Nice! great way to reduce 8 lines to 2 _and_ no more having to add new lines for each module! 👍

File: docs/source/contribution-guidelines.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (WiscEvan) [reply docs/source/contribution-guidelines.rst:6]: I supposed our separate TODO page is our list of issues on GH right now. This may be related or unrelated, but as we release each new tagged version, there should be a changes page describing the features added, bugfixes etc. Although, I think this may be outside of the scope of this PR.

File: docs/source/examining-results.rst [hunk 12]
```diff
@@ -3,7 +3,7 @@ Examining Results
 =================
 
 
-TODO: Overview of figures/summaries generated from results.
+.. todo:: Overview of figures/summaries generated from results.
```
Comment (WiscEvan) [inline docs/source/examining-results.rst:6]: This looks so nice. Thanks for adding the extension 👍

File: docs/source/index.rst [hunk 13]
```diff
@@ -38,11 +30,16 @@ Guide
...
+
 Contact
 -------
 For suggestions, queries or appreciation feel free to contact `Dr.JasonKwan <https://apps.pharmacy.wisc.edu/sopdir/jason_kwan/>`_
 at jason.kwan@wisc.edu
 
+Copyright and License
+---------------------
+Autometa is copyright 2020 Ian J. Miller, Evan R. Rees, Izaak Miller and Jason C. Kwan, 
+and is released under the GNU Affero General Public License v3 (see :ref:`License`).
```
Comment (WiscEvan) [inline docs/source/index.rst:42]: So copyright is also in the footer of each page. I'm not sure it is necessary at this point to be hardcoded in here. Note, Ian and Izaak will need to be added to the other copyright section.

File: docs/source/Scripts/Common/External/bowtie.rst [hunk 14]
```diff
@@ -0,0 +1,19 @@
+=========
+bowtie.py
+=========
+
+.. rubric:: Usage adn command line options
```
Comment (WiscEvan) [inline docs/source/Scripts/Common/External/bowtie.rst:5]: Typo with `adn` Usage `and` command line options

File: docs/source/_ext/todo.py (hunk 10 above)
Comment (Sidduppal) [reply docs/source/_ext/todo.py:127]: I did not write it, it was taken from here https://www.sphinx-doc.org/en/master/development/tutorials/todo.html.
Any particular format to reference the source?

File: docs/source/Scripts/Common/External/bowtie.rst (hunk 14 above)
Comment (Sidduppal) [reply docs/source/Scripts/Common/External/bowtie.rst:5]: Thanks for catching that

File: docs/source/_ext/todo.py (hunk 10 above)
Comment (Sidduppal) [reply docs/source/_ext/todo.py:127]: There are a couple of sphinx documentation that tell you how to build extensions. Fortunately, they had one for making a `todo` extension.

Comment (WiscEvan) [reply docs/source/_ext/todo.py]: You should at least place this as a comment in `todo.py`

File: docs/source/index.rst [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+   api
+   license
+
+
+Contact
+-------
+For suggestions, queries or appreciation feel free to contact `Dr.JasonKwan <https://apps.pharmacy.wisc.edu/sopdir/jason_kwan/>`_
+at jason.kwan@wisc.edu
+
+Copyright and License
```
Comment (WiscEvan) [inline docs/source/index.rst:39]: You could probably just remove this section Since License is in the sidebar. At least remove the Copyright portion
